### PR TITLE
IGNITE-17555 Fix AssertionError and NPE in configuration

### PR DESCRIPTION
**Pull request by SammyVimes**
https://issues.apache.org/jira/browse/IGNITE-17555

### CODE REVIEW

File: modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java [hunk 1]
```diff
@@ -247,23 +266,6 @@ public CompletableFuture<Boolean> write(Map<String, ? extends Serializable> newV
...
-        //       observed any configuration changes. Valid condition is "MASTER_KEY does not exist".
-        //     - MASTER_KEY is present in local MetaStorage copy. The MASTER_KEY revision is unknown but is less than or
-        //       equal to APPLIED_REV. Obviously, there have been no updates from the future yet. It's also guaranteed
-        //       that the next received configuration update will have the MASTER_KEY revision strictly greater than
-        //       current APPLIED_REV. This allows to conclude that "MASTER_KEY revision <= curChangeId" is a valid
-        //       condition for update.
-        // Combining all of the above, it's concluded that the following condition must be used:
         SimpleCondition condition = curChangeId == 0L
                 ? Conditions.notExists(MASTER_KEY)
                 : Conditions.revision(MASTER_KEY).le(curChangeId);
```
Comment (ibessonov) [inline modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java:271]: Should replace that with `eq` probably

Comment (SammyVimes) [reply modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java:271]: oh, right

Comment (SammyVimes) [reply modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java]: fixed

File: modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java [hunk 2]
```diff
@@ -74,6 +93,145 @@ void stop() throws Exception {
...
+            assertEquals(configurationChangesCount, lastConfigurationChangeRevision);
+
+            changer.stop();
+        }
+    }
+
+    /**
+     * This class stores data for {@link MetaStorageManager}'s mock.
+     */
+    private static class MetaStorageMockWrapper {
```
Comment (sashapolo) [inline modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java:167]: we now have two mock metastorage managers here. I think this test should be extracted into a separate class

File: modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java [hunk 3]
```diff
@@ -74,6 +93,145 @@ void stop() throws Exception {
...
+            changer.start();
+
+            CompletableFuture<Long> longCompletableFuture = storage.lastRevision();
+
+            // Should return last configuration change, not last MetaStorage change.
+            Long lastConfigurationChangeRevision = longCompletableFuture.get();
+
+            assertEquals(configurationChangesCount, lastConfigurationChangeRevision);
+
+            changer.stop();
```
Comment (sashapolo) [inline modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java:160]: I guess this should be moved to a `finally` block

File: modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java [hunk 4]
```diff
@@ -74,6 +93,145 @@ void stop() throws Exception {
...
+     */
+    @Test
+    public void testMetaStorageRevisionDifferentFromConfigurationOnRestart() throws Exception {
+        RootKey<DistributedTestConfiguration, DistributedTestView> rootKey = DistributedTestConfiguration.KEY;
+
+        ConfigurationAsmGenerator cgen = new ConfigurationAsmGenerator();
+
+        MetaStorageMockWrapper wrapper = new MetaStorageMockWrapper();
+
+        int configurationChangesCount = 7;
```
Comment (sashapolo) [inline modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java:119]: why 7? Isn't one change enough?

File: modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java [hunk 5]
```diff
@@ -74,6 +93,145 @@ void stop() throws Exception {
...
+        // This emulates a change in MetaStorage that is not related to the configuration.
+        vaultManager.put(MetaStorageManager.APPLIED_REV, ByteUtils.longToBytes(configurationChangesCount + 1)).get();
+
+        try (var storage = new DistributedConfigurationStorage(wrapper.metaStorageManager(), vaultManager)) {
+            var changer = new TestConfigurationChanger(cgen, List.of(rootKey), Collections.emptyMap(),
+                    storage, Collections.emptyList(), Collections.emptyList());
+
+            changer.start();
+
+            CompletableFuture<Long> longCompletableFuture = storage.lastRevision();
```
Comment (sashapolo) [inline modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java:153]: I guess this future can be inlined

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/storage/ConfigurationStorage.java [hunk 6]
```diff
@@ -27,11 +27,11 @@
  */
 public interface ConfigurationStorage extends AutoCloseable {
     /**
-     * Read all configuration values and current storage version.
+     * Reads all configuration values and current storage version on the start.
```
[the remark this answers is not in the document]
Comment (SammyVimes) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/storage/ConfigurationStorage.java:30]: As I understand, it only makes sense on the start (and it's only called on the start). It does some additional stuff like deciding what revision to use

File: modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java [hunk 7]
```diff
@@ -86,7 +86,7 @@ public static byte[] longToBytes(long l) {
      * @param limit Limit of bytes to write into output.
      * @return Number of bytes overwritten in {@code bytes} array.
      */
-    private static byte[] longToBytes(long l, byte[] bytes, int off, int limit) {
+    public static byte[] longToBytes(long l, byte[] bytes, int off, int limit) {
```
[the remark this answers is not in the document]
Comment (SammyVimes) [reply modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java:89]: You're right

File: modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java [hunk 8]
```diff
@@ -247,23 +266,6 @@ public CompletableFuture<Boolean> write(Map<String, ? extends Serializable> newV
 
         operations.add(Operations.put(MASTER_KEY, ByteUtils.longToBytes(curChangeId)));
 
-        // Condition for a valid MetaStorage data update. Several possibilities here:
```
[the remark this answers is not in the document]
Comment (SammyVimes) [reply modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java:250]: The original author of the comment (Ivan) suggested that it is not needed anymore as we have a more strict contract on what the value might be

File: modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java [hunk 9]
```diff
@@ -185,38 +190,52 @@ public CompletableFuture<Serializable> readLatest(String key) {
 
     /** {@inheritDoc} */
     @Override
-    public CompletableFuture<Data> readAll() throws StorageException {
+    public CompletableFuture<Data> readAllOnStart() throws StorageException {
         return registerFuture(vaultMgr.get(MetaStorageManager.APPLIED_REV)
-                .thenApplyAsync(appliedRevEntry -> {
-                    long appliedRevision = appliedRevEntry == null ? 0L : ByteUtils.bytesToLong(appliedRevEntry.value());
+                .thenComposeAsync(appliedRevEntry -> {
```
[the remark this answers is not in the document]
Comment (SammyVimes) [reply modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java:195]: I knew you'd be able to suggest something about this mess!

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/storage/ConfigurationStorage.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/storage/ConfigurationStorage.java:30]: Maybe we shall rename it to something like "performRecovery"?

File: modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java [hunk 10]
```diff
@@ -74,6 +93,145 @@ void stop() throws Exception {
...
+
+    /**
+     * This class stores data for {@link MetaStorageManager}'s mock.
+     */
+    private static class MetaStorageMockWrapper {
+        private static final String DISTRIBUTED_PREFIX = "dst-cfg.";
+
+        /**
+         * This and previous field are copy-pasted intentionally, so in case if something changes,
+         * this test should fail and be reviewed and re-wrote.
```
[the remark this answers is not in the document]
Comment (SammyVimes) [reply modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java:172]: whoops

File: modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java:195]: Or we can take this even further and split `readAll` into two methods: one to resolve the required revision and another to read the data

File: modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java:250]: But I suppose that some parts of this comment are still valid, aren't they?

File: modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java (hunk 2 above)
Comment (SammyVimes) [reply modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java:167]: right

File: modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java (hunk 4 above)
Comment (SammyVimes) [reply modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java:119]: It's a tribute to the original bug I was constantly facing. It's extremely fast, so I feel like it's ok? Or I can change it if it looks cryptic

Comment (sashapolo) [reply modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java:119]: It looks cryptic indeed, can you add a comment?

File: modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/EntryImpl.java [hunk 11]
```diff
@@ -48,7 +48,7 @@ public final class EntryImpl implements Entry {
      * @param rev     Revision.
      * @param updCntr Update counter.
      */
-    EntryImpl(@NotNull ByteArray key, @Nullable byte[] val, long rev, long updCntr) {
+    public EntryImpl(@NotNull ByteArray key, @Nullable byte[] val, long rev, long updCntr) {
```
[the remark this answers is not in the document]
Comment (SammyVimes) [reply modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/EntryImpl.java:51]: Sure

File: modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (SammyVimes) [reply modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java:250]: Yes, but only the obvious part, so "we only can change metastorage if didn't change"

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/storage/ConfigurationStorage.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/storage/ConfigurationStorage.java:30]: readDataOnRecovery is also fine, I guess

File: modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java:250]: ok, let's remove it then

File: modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java (hunk 4 above)
Comment (SammyVimes) [reply modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorageTest.java:119]: Sure

File: modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationCatchUpTest.java [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+     */
+    @Test
+    public void testMetaStorageRevisionDifferentFromConfigurationOnRestart() throws Exception {
+        RootKey<DistributedTestConfiguration, DistributedTestView> rootKey = DistributedTestConfiguration.KEY;
+
+        ConfigurationAsmGenerator cgen = new ConfigurationAsmGenerator();
+
+        MetaStorageMockWrapper wrapper = new MetaStorageMockWrapper();
+
+        // This value is a tribute to the specific error that was occurring. Can as well be 1.
```
Comment (sashapolo) [inline modules/runner/src/test/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationCatchUpTest.java:97]: This comment is as cryptic as the value 7 =)

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/ConfigurationChanger.java [hunk 13]
```diff
@@ -572,20 +572,24 @@ private CompletableFuture<Void> updateFromListener(Data changedEntries) {
...
+        return storage.writeConfigurationRevision(oldStorageRoots.version, storageRoots.version)
+                .thenCompose(unused -> {
+                    if (dataValuesPrefixMap.isEmpty()) {
                         oldStorageRoots.changeFuture.complete(null);
+
+                        return CompletableFuture.completedFuture(null);
                     } else {
-                        oldStorageRoots.changeFuture.completeExceptionally(t);
+                        long notificationNumber = notificationListenerCnt.incrementAndGet();
+                        return notificator.notify(oldSuperRoot, newSuperRoot, newChangeId, notificationNumber).whenComplete((v, t) -> {
```
Comment (sashapolo) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/ConfigurationChanger.java:584]: 1. There should be a line break.
2. The same thing: I would suggest to move `whenComplete` to the next line

File: modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java [hunk 14]
```diff
@@ -185,39 +190,54 @@ public CompletableFuture<Serializable> readLatest(String key) {
...
-    public CompletableFuture<Data> readAll() throws StorageException {
+    public CompletableFuture<Data> readDataOnRecovery() throws StorageException {
         return registerFuture(vaultMgr.get(MetaStorageManager.APPLIED_REV)
-                .thenApplyAsync(appliedRevEntry -> {
-                    long appliedRevision = appliedRevEntry == null ? 0L : ByteUtils.bytesToLong(appliedRevEntry.value());
+                .thenCombineAsync(vaultMgr.get(CONFIGURATION_REVISIONS_KEY), this::readAllOnStart0, threadPool));
+    }
+
+    @NotNull
+    private Data readAllOnStart0(VaultEntry appliedRevEntry, VaultEntry revisionsEntry) {
```
Comment (sashapolo) [inline modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/DistributedConfigurationStorage.java:199]: 1. Parameters should be marked with `@Nullable`
2. We don't use `@NotNull`

**Pull request merged**